File: app/src/main/java/com/ostrava/app/domain/Calculations.kt
package com.ostrava.app.domain

import android.location.Location

private fun distanceBetween(a: TrackPoint, b: TrackPoint): Double {
    val results = FloatArray(1)
    Location.distanceBetween(a.latitude, a.longitude, b.latitude, b.longitude, results)
    return results[0].toDouble()
}

private fun distanceToPoint(lat: Double, lon: Double, point: TrackPoint): Double {
    val results = FloatArray(1)
    Location.distanceBetween(lat, lon, point.latitude, point.longitude, results)
    return results[0].toDouble()
}

/**
 * Buckets the track into consecutive splits of [splitLengthMeters], linearly interpolating
 * the boundary crossing time so split durations are accurate.
 */
fun computeSplits(points: List<TrackPoint>, splitLengthMeters: Double): List<Split> {
    if (points.size < 2) return emptyList()
    val splits = mutableListOf<Split>()
    var splitDistance = 0.0
    var splitStartTime = points.first().timeMillis
    var splitStartAltitude = points.first().altitude
    var index = 1

    var prev = points.first()
    for (point in points.drop(1)) {
        if (point.segment != prev.segment) {
            // Across a pause: no distance accrues and time gap is skipped by shifting the split start.
            splitStartTime += point.timeMillis - prev.timeMillis
            prev = point
            continue
        }
        var legDistance = distanceBetween(prev, point)
        var legStartTime = prev.timeMillis
        var legDuration = (point.timeMillis - prev.timeMillis).toDouble()

        while (splitDistance + legDistance >= splitLengthMeters && legDistance > 0) {
            val needed = splitLengthMeters - splitDistance
            val fraction = needed / legDistance
            val crossingTime = legStartTime + (legDuration * fraction).toLong()
            splits += Split(
                index = index,
                distanceMeters = splitLengthMeters,
                durationMillis = crossingTime - splitStartTime,
                elevationDelta = point.altitude - splitStartAltitude,
            )
            index++
            splitStartTime = crossingTime
            splitStartAltitude = point.altitude
            legDuration -= legDuration * fraction
            legStartTime = crossingTime
            legDistance -= needed
            splitDistance = 0.0
        }
        splitDistance += legDistance
        prev = point
    }
    if (splitDistance > splitLengthMeters * 0.05) {
        splits += Split(
            index = index,
            distanceMeters = splitDistance,
            durationMillis = points.last().timeMillis - splitStartTime,
            elevationDelta = points.last().altitude - splitStartAltitude,
        )
    }
    return splits
}

/**
 * Fastest contiguous effort covering [targetMeters], found with a two-pointer sweep over
 * cumulative distance/time. Returns null if the activity is shorter than the target.
 */
fun bestEffort(points: List<TrackPoint>, targetMeters: Double, label: String): BestEffort? {
    if (points.size < 2) return null
    val n = points.size
    val cumDistance = DoubleArray(n)
    val cumTime = LongArray(n)
    for (i in 1 until n) {
        val sameSegment = points[i].segment == points[i - 1].segment
        cumDistance[i] = cumDistance[i - 1] +
            if (sameSegment) distanceBetween(points[i - 1], points[i]) else 0.0
        cumTime[i] = cumTime[i - 1] +
            if (sameSegment) points[i].timeMillis - points[i - 1].timeMillis else 0L
    }
    if (cumDistance[n - 1] < targetMeters) return null

    var best = Long.MAX_VALUE
    var start = 0
    for (end in 1 until n) {
        while (cumDistance[end] - cumDistance[start + 1] >= targetMeters && start + 1 < end) {
            start++
        }
        if (cumDistance[end] - cumDistance[start] >= targetMeters) {
            val duration = cumTime[end] - cumTime[start]
            if (duration in 1 until best) best = duration
        }
    }
    return if (best == Long.MAX_VALUE) null else BestEffort(label, targetMeters, best)
}

fun bestEffortsFor(type: ActivityType, points: List<TrackPoint>): List<BestEffort> {
    val targets = when (type) {
        ActivityType.RUN -> listOf("1 km" to 1000.0, "5 km" to 5000.0, "10 km" to 10000.0, "Half marathon" to 21097.5)
        ActivityType.RIDE -> listOf("5 km" to 5000.0, "20 km" to 20000.0, "40 km" to 40000.0)
        else -> listOf("1 km" to 1000.0, "5 km" to 5000.0)
    }
    return targets.mapNotNull { (label, meters) -> bestEffort(points, meters, label) }
}

/** MET-based calorie estimate. */
fun estimateCalories(type: ActivityType, avgSpeedMps: Double, movingTimeMillis: Long, weightKg: Float): Int {
    val kmh = avgSpeedMps * 3.6
    val met = when (type) {
        ActivityType.RUN -> maxOf(6.0, kmh * 1.0)
        ActivityType.RIDE -> when {
            kmh < 16 -> 6.0
            kmh < 20 -> 8.0
            kmh < 25 -> 10.0
            else -> 12.0
        }
        ActivityType.WALK -> 3.5
        ActivityType.HIKE -> 6.0
    }
    val hours = movingTimeMillis / 3_600_000.0
    return (met * weightKg * hours).toInt()
}

/** Summary stats derived from a raw track, used by GPX import and backup restore. */
data class ComputedStats(
    val distanceMeters: Double,
    val movingTimeMillis: Long,
    val elevationGainMeters: Double,
    val avgSpeedMps: Double,
    val maxSpeedMps: Double,
)

fun computeActivityStats(points: List<TrackPoint>): ComputedStats {
    if (points.size < 2) return ComputedStats(0.0, 0L, 0.0, 0.0, 0.0)
    var distance = 0.0
    var movingTime = 0L
    var gain = 0.0
    var maxSpeed = 0.0
    var altitudeBaseline = points.first().altitude

    for (i in 1 until points.size) {
        val prev = points[i - 1]
        val point = points[i]
        if (point.segment != prev.segment) {
            altitudeBaseline = point.altitude
            continue
        }
        val d = distanceBetween(prev, point)
        val dt = point.timeMillis - prev.timeMillis
        distance += d
        // Gaps longer than 15 s are treated as standing still, not moving time.
        if (dt in 1..15_000) {
            movingTime += dt
            val legSpeed = d / (dt / 1000.0)
            if (legSpeed > maxSpeed && legSpeed < 40.0) maxSpeed = legSpeed
        }
        when {
            point.altitude - altitudeBaseline >= 2.0 -> {
                gain += point.altitude - altitudeBaseline
                altitudeBaseline = point.altitude
            }
            point.altitude < altitudeBaseline -> altitudeBaseline = point.altitude
        }
    }
    val avgSpeed = if (movingTime > 0) distance / (movingTime / 1000.0) else 0.0
    return ComputedStats(distance, movingTime, gain, avgSpeed, maxSpeed)
}

/** A matched segment traversal within an activity. */
data class SegmentMatch(
    val durationMillis: Long,
    val startTimeMillis: Long,
)

private const val SEGMENT_MATCH_RADIUS_METERS = 40.0

/**
 * Finds the fastest traversal of a segment (start gate -> end gate) within a track.
 * A traversal counts when the path passes within [SEGMENT_MATCH_RADIUS_METERS] of both
 * gates and covers a path length comparable to the segment's distance.
 */
fun matchSegment(
    startLat: Double,
    startLon: Double,
    endLat: Double,
    endLon: Double,
    segmentDistanceMeters: Double,
    points: List<TrackPoint>,
): SegmentMatch? {
    if (points.size < 2 || segmentDistanceMeters <= 0) return null
    val n = points.size
    val cumDistance = DoubleArray(n)
    val cumTime = LongArray(n)
    for (i in 1 until n) {
        val sameSegment = points[i].segment == points[i - 1].segment
        cumDistance[i] = cumDistance[i - 1] +
            if (sameSegment) distanceBetween(points[i - 1], points[i]) else 0.0
        cumTime[i] = cumTime[i - 1] +
            if (sameSegment) points[i].timeMillis - points[i - 1].timeMillis else 0L
    }

    val minPath = segmentDistanceMeters * 0.75
    val maxPath = segmentDistanceMeters * 1.35
    var best: SegmentMatch? = null

    var i = 0
    while (i < n) {
        if (distanceToPoint(startLat, startLon, points[i]) <= SEGMENT_MATCH_RADIUS_METERS) {
            var j = i + 1
            while (j < n && cumDistance[j] - cumDistance[i] <= maxPath) {
                if (cumDistance[j] - cumDistance[i] >= minPath &&
                    distanceToPoint(endLat, endLon, points[j]) <= SEGMENT_MATCH_RADIUS_METERS
                ) {
                    val duration = cumTime[j] - cumTime[i]
                    if (duration > 0 && (best == null || duration < best.durationMillis)) {
                        best = SegmentMatch(duration, points[i].timeMillis)
                    }
                    break
                }
                j++
            }
            // Skip past this start-gate cluster before looking for another traversal.
            while (i + 1 < n &&
                distanceToPoint(startLat, startLon, points[i + 1]) <= SEGMENT_MATCH_RADIUS_METERS
            ) {
                i++
            }
        }
        i++
    }
    return best
}

/**
 * Weekly streaks from the set of week indices that contain at least one activity.
 * Returns (current, longest). The current streak tolerates the present week being
 * empty so a Monday doesn't zero everyone's streak.
 */
fun weeklyStreaks(activeWeekIndices: Set<Long>, currentWeekIndex: Long): Pair<Int, Int> {
    if (activeWeekIndices.isEmpty()) return 0 to 0

    var current = 0
    var week = if (currentWeekIndex in activeWeekIndices) currentWeekIndex else currentWeekIndex - 1
    while (week in activeWeekIndices) {
        current++
        week--
    }

    var longest = 0
    var run = 0
    var previous: Long? = null
    for (w in activeWeekIndices.sorted()) {
        run = if (previous != null && w == previous + 1) run + 1 else 1
        if (run > longest) longest = run
        previous = w
    }
    return current to longest
}

File: ios/Ostrava/Calculations.swift
import CoreLocation
import Foundation

private func distanceMeters(_ a: TrackPoint, _ b: TrackPoint) -> Double {
    CLLocation(latitude: a.latitude, longitude: a.longitude)
        .distance(from: CLLocation(latitude: b.latitude, longitude: b.longitude))
}

private func distanceToPoint(lat: Double, lon: Double, point: TrackPoint) -> Double {
    CLLocation(latitude: lat, longitude: lon)
        .distance(from: CLLocation(latitude: point.latitude, longitude: point.longitude))
}

struct Split: Identifiable {
    var index: Int
    var distance: Double
    var duration: TimeInterval
    var elevationDelta: Double
    var id: Int { index }

    var paceSecondsPerKm: Double {
        distance > 0 ? duration / (distance / 1000.0) : 0
    }
}

/// Consecutive splits of `splitLength` meters, interpolating boundary crossing time.
func computeSplits(points: [TrackPoint], splitLength: Double) -> [Split] {
    guard points.count >= 2 else { return [] }
    var splits: [Split] = []
    var splitDistance = 0.0
    var splitStartTime = points[0].time
    var splitStartAltitude = points[0].altitude
    var index = 1

    var prev = points[0]
    for point in points.dropFirst() {
        if point.segment != prev.segment {
            splitStartTime += point.time - prev.time
            prev = point
            continue
        }
        var legDistance = distanceMeters(prev, point)
        var legStartTime = prev.time
        var legDuration = point.time - prev.time

        while splitDistance + legDistance >= splitLength && legDistance > 0 {
            let needed = splitLength - splitDistance
            let fraction = needed / legDistance
            let crossingTime = legStartTime + legDuration * fraction
            splits.append(
                Split(
                    index: index,
                    distance: splitLength,
                    duration: crossingTime - splitStartTime,
                    elevationDelta: point.altitude - splitStartAltitude
                )
            )
            index += 1
            splitStartTime = crossingTime
            splitStartAltitude = point.altitude
            legDuration -= legDuration * fraction
            legStartTime = crossingTime
            legDistance -= needed
            splitDistance = 0
        }
        splitDistance += legDistance
        prev = point
    }
    if splitDistance > splitLength * 0.05 {
        splits.append(
            Split(
                index: index,
                distance: splitDistance,
                duration: points[points.count - 1].time - splitStartTime,
                elevationDelta: points[points.count - 1].altitude - splitStartAltitude
            )
        )
    }
    return splits
}

struct BestEffortResult: Identifiable {
    var label: String
    var distance: Double
    var duration: TimeInterval
    var id: String { label }
}

/// Fastest contiguous stretch covering `target` meters (two-pointer sweep).
func bestEffort(points: [TrackPoint], target: Double, label: String) -> BestEffortResult? {
    guard points.count >= 2 else { return nil }
    let n = points.count
    var cumDistance = [Double](repeating: 0, count: n)
    var cumTime = [Double](repeating: 0, count: n)
    for i in 1..<n {
        let same = points[i].segment == points[i - 1].segment
        cumDistance[i] = cumDistance[i - 1] + (same ? distanceMeters(points[i - 1], points[i]) : 0)
        cumTime[i] = cumTime[i - 1] + (same ? points[i].time - points[i - 1].time : 0)
    }
    guard cumDistance[n - 1] >= target else { return nil }

    var best = Double.greatestFiniteMagnitude
    var start = 0
    for end in 1..<n {
        while start + 1 < end && cumDistance[end] - cumDistance[start + 1] >= target {
            start += 1
        }
        if cumDistance[end] - cumDistance[start] >= target {
            let duration = cumTime[end] - cumTime[start]
            if duration > 0 && duration < best { best = duration }
        }
    }
    return best == .greatestFiniteMagnitude ? nil
        : BestEffortResult(label: label, distance: target, duration: best)
}

func bestEffortsFor(type: ActivityType, points: [TrackPoint]) -> [BestEffortResult] {
    let targets: [(String, Double)]
    switch type {
    case .run:
        targets = [("1 km", 1000), ("5 km", 5000), ("10 km", 10000), ("Half marathon", 21097.5)]
    case .ride:
        targets = [("5 km", 5000), ("20 km", 20000), ("40 km", 40000)]
    default:
        targets = [("1 km", 1000), ("5 km", 5000)]
    }
    return targets.compactMap { bestEffort(points: points, target: $0.1, label: $0.0) }
}

/// MET-based calorie estimate.
func estimateCalories(type: ActivityType, avgSpeedMps: Double, movingTime: TimeInterval, weightKg: Double) -> Int {
    let kmh = avgSpeedMps * 3.6
    let met: Double
    switch type {
    case .run: met = max(6.0, kmh)
    case .ride:
        if kmh < 16 { met = 6 } else if kmh < 20 { met = 8 } else if kmh < 25 { met = 10 } else { met = 12 }
    case .walk: met = 3.5
    case .hike: met = 6.0
    }
    return Int(met * weightKg * movingTime / 3600.0)
}

struct ComputedStats {
    var distance: Double
    var movingTime: TimeInterval
    var elevationGain: Double
    var avgSpeed: Double
    var maxSpeed: Double
}

/// Summary stats from a raw track (GPX import, backup restore).
func computeActivityStats(points: [TrackPoint]) -> ComputedStats {
    guard points.count >= 2 else {
        return ComputedStats(distance: 0, movingTime: 0, elevationGain: 0, avgSpeed: 0, maxSpeed: 0)
    }
    var distance = 0.0
    var movingTime = 0.0
    var gain = 0.0
    var maxSpeed = 0.0
    var baseline = points[0].altitude

    for i in 1..<points.count {
        let prev = points[i - 1]
        let point = points[i]
        if point.segment != prev.segment {
            baseline = point.altitude
            continue
        }
        let d = distanceMeters(prev, point)
        let dt = point.time - prev.time
        distance += d
        if dt > 0 && dt <= 15 {
            movingTime += dt
            let legSpeed = d / dt
            if legSpeed > maxSpeed && legSpeed < 40 { maxSpeed = legSpeed }
        }
        if point.altitude - baseline >= 2 {
            gain += point.altitude - baseline
            baseline = point.altitude
        } else if point.altitude < baseline {
            baseline = point.altitude
        }
    }
    let avgSpeed = movingTime > 0 ? distance / movingTime : 0
    return ComputedStats(
        distance: distance, movingTime: movingTime, elevationGain: gain,
        avgSpeed: avgSpeed, maxSpeed: maxSpeed
    )
}

struct SegmentMatch {
    var duration: TimeInterval
    var startTime: TimeInterval
}

private let segmentMatchRadius = 40.0

/// Fastest traversal of a start->end gate pair within a track, if any.
func matchSegment(
    startLat: Double, startLon: Double,
    endLat: Double, endLon: Double,
    segmentDistance: Double,
    points: [TrackPoint]
) -> SegmentMatch? {
    guard points.count >= 2, segmentDistance > 0 else { return nil }
    let n = points.count
    var cumDistance = [Double](repeating: 0, count: n)
    var cumTime = [Double](repeating: 0, count: n)
    for i in 1..<n {
        let same = points[i].segment == points[i - 1].segment
        cumDistance[i] = cumDistance[i - 1] + (same ? distanceMeters(points[i - 1], points[i]) : 0)
        cumTime[i] = cumTime[i - 1] + (same ? points[i].time - points[i - 1].time : 0)
    }

    let minPath = segmentDistance * 0.75
    let maxPath = segmentDistance * 1.35
    var best: SegmentMatch?

    var i = 0
    while i < n {
        if distanceToPoint(lat: startLat, lon: startLon, point: points[i]) <= segmentMatchRadius {
            var j = i + 1
            while j < n && cumDistance[j] - cumDistance[i] <= maxPath {
                if cumDistance[j] - cumDistance[i] >= minPath
                    && distanceToPoint(lat: endLat, lon: endLon, point: points[j]) <= segmentMatchRadius {
                    let duration = cumTime[j] - cumTime[i]
                    if duration > 0 && (best == nil || duration < best!.duration) {
                        best = SegmentMatch(duration: duration, startTime: points[i].time)
                    }
                    break
                }
                j += 1
            }
            while i + 1 < n
                && distanceToPoint(lat: startLat, lon: startLon, point: points[i + 1]) <= segmentMatchRadius {
                i += 1
            }
        }
        i += 1
    }
    return best
}

/// (current, longest) weekly streaks; the present week may still be empty.
func weeklyStreaks(activeWeekIndices: Set<Int>, currentWeekIndex: Int) -> (current: Int, longest: Int) {
    guard !activeWeekIndices.isEmpty else { return (0, 0) }

    var current = 0
    var week = activeWeekIndices.contains(currentWeekIndex) ? currentWeekIndex : currentWeekIndex - 1
    while activeWeekIndices.contains(week) {
        current += 1
        week -= 1
    }

    var longest = 0
    var run = 0
    var previous: Int?
    for w in activeWeekIndices.sorted() {
        run = (previous != nil && w == previous! + 1) ? run + 1 : 1
        longest = max(longest, run)
        previous = w
    }
    return (current, longest)
}

/// Monday-based week index for streak/load bucketing.
func weekIndex(of date: Date) -> Int {
    var calendar = Calendar(identifier: .iso8601)
    calendar.firstWeekday = 2
    let monday = calendar.dateInterval(of: .weekOfYear, for: date)?.start ?? date
    return Int(monday.timeIntervalSince1970 / 604_800)
}
